pkg/controller/mongodb/mongodb_controller.go

```diff
@@ -291,24 +336,75 @@ func (r ReplicaSetReconciler) ensureAutomationConfig(mdb mdbv1.MongoDB) error {
 	return nil
 }
 
-func buildAutomationConfig(mdb mdbv1.MongoDB, mdbVersionConfig automationconfig.MongoDbVersionConfig, client mdbClient.Client) (automationconfig.AutomationConfig, error) {
-	domain := getDomain(mdb.ServiceName(), mdb.Namespace, "")
+// checkTLSConfig will check that the configured ConfigMap and Secret exist and that they have the correct fields.
```

Comment:
The docs are a bit misleading, the first return value is a "shouldRetry", but I had to go and look at the usages to understand it.

Returning a "true" value from a `checkTLSConfig` should indicate that *everything is ok*, but in this case, it means that the *process should retry*, which is kind of confusing.

At the same time, I think that both values can be wrapped into the same error type: if the function returns err == nil, we can proceed. Reconcile otherwise. <- this can be discussed

Reply:
I think this function is creating a lot of confusion and should be handled differently

maybe using a decorator pattern that handles the retry will make it clearer

Reply:
I think you're right, the returned boolean is confusing. Positive bool should definitely indicate a proper TLS config.

Regarding just returning an error, I separated these because there are two error cases:
1. The user hasn't set up the required secret or configmap correctly. In this case we want to retry in a little while to give the user time to fix it.
2. Something fails when checking the resources, for example a intermittent error. In this case we want to retry immediately.

1. could of course be implemented by returning an error as well but that would cause an immediate requeue and the operator looping many times before the user has a chance to update the configuration.

Reply:
@LouisPlisso your comments basically come down to this as well. Would love to hear your take on whether we should do an immediate requeue, and if not, how to best handle this three case situation.

Reply:
see comment below: a retryable error should not be surfaced as an error
if err is not nil, then you can use the retryable value

if I understand the code correctly, right now we have these possible outputs: (false, nil), (true, err), (false, err)

golang pattern would be to have: (false, nil), (true, nil), (false, err) [false with err because it's the default value of bool, but it should never be used if err is non-nil)

does that help?

Reply:
Right, I agree! I'm not sure how to surface the feedback message for the user though in case they have something wrong in their configuration.

Basically right now I'm encoding both user errors and other errors in the return `error` which is dumb as it's not idiomatic Go. Using `(false, nil)`, `(true, nil)`, `(false, err)` wouldn't make it possible to return an error message for the user to see what is wrong with their config.

I have a few ideas but I'm not sure which one is the most idiomatic:
1. Return three values: `(valid bool, validationMessage string, err error)`. If `valid` is true then the reconcilation proceeds, if `err` is not nil then two others are ignored, and if `valid` is false then `validationMessage` contains some feedback for the user.
2. Encode the validation result in a new type `TLSValidationResult` and return `(TLSValidationResult, error)`. This struct could contain two values: `Valid bool` and `ValidationMessage string`. This is similar to solution 1.
3. Create a new error type for user config errors and then have a helper method that can check if an error is a user error (similarly to the Kubernetes error helpers). This way the function would only have to return a single `error`. This is my least favourite as the different cases get lost in the type system.

Reply:
@fabianlindfors This is something has come up a lot, we want to represent more than two possible outcomes, but the restriction of `reconcilliation.Result, error` doesn't allow us to adequately represent the various states we encounter. I encourage you to take a look at the [workflow](https://github.com/10gen/ops-manager-kubernetes/blob/3877ddc263ec5576d8ceb98003ee35e819dbac7e/pkg/controller/operator/workflow/status.go#L13) package we have in enterprise (not to do a direct port, but to maybe get some ideas from it)

I think we will need to introduce something similar to these to account for all the states we will encounter.

Reply:
I think the type of refactor I suggested is something for a future PR, of your suggestions, I like number 2 the most, with the intention of creating something that we can generally use to express different types of results.

Reply:
Sounds like a good plan! I've created an internal ticket CLOUDP-66375 to create something similar to the workflow package for community.

I'll go with number 2 for now which should improve readability.

Reply:
I've now implemented option 2 with a type called `tlsValidationResult`. This means that the function, now called `validateTLSConfig`, returns `(tlsValidationResult, error)`. `error` is only set if there was an actual error during validation and not for user errors.

CLOUDP-66375 will replace this with a more general solution for handling multi-case reconciliation results.

Reply:
@rodrigovalin do you like the new interface for `validateTLSConfig` more?